shesmu-server-ui/src/html.ts

```diff
@@ -0,0 +1,1618 @@
+}
+
+/**
+ * Add a blank UI element
+ *
+ * This isn't very useful, but simplifies some code paths in the action tile building.
+ */
+export function blank(): UIElement {
+  return [];
+}
+/**
+ * Create a line break
+ */
+export function br(): UIElement {
+  return document.createElement("br");
+}
+
+/**
+ * Create a modal dialog with a throbber in the middle.
+ *
+ * It returns a function to close this dialog
+ */
+export function busyDialog(): () => void {
+  const modal = document.createElement("div");
+  modal.className = "modal";
+  addElements(modal, throbber());
+  document.body.appendChild(modal);
+  return () => document.body.removeChild(modal);
+}
+
+/**
+ * Create a normal button
+ */
+export function button(
+  label: UIElement,
+  title: string,
+  callback: ClickHandler
+) {
+  return buttonCustom(label, title, [], callback);
+}
+
+/**
+ * Create a button for an unimportant feature
+ */
+export function buttonAccessory(
+  label: UIElement,
+  title: string,
+  callback: ClickHandler
+) {
+  return buttonCustom(label, title, ["accessory"], callback);
+}
+
+/**
+ * Create a button for with a custom UI design
+ */
+export function buttonCustom(
+  label: UIElement,
+  title: string,
+  className: string[],
+  callback: ClickHandler
+) {
+  const button = document.createElement("span");
+  button.classList.add("load");
+  for (const name of className) {
+    button.classList.add(name);
+  }
+  addElements(button, label);
+  button.title = title;
+  button.addEventListener("click", callback);
+  return button;
+}
+
+/**
+ * Create a close button
+ */
+export function buttonClose(title: string, callback: ClickHandler): UIElement {
+  return buttonSubtle("✖", title, callback);
+}
+/**
+ * Create a button for a dangerous feature
+ */
+export function buttonDanger(
+  label: UIElement,
+  title: string,
+  callback: ClickHandler
+) {
+  return buttonCustom(label, title, ["danger"], callback);
+}
+/**
+ * Create an edit button
+ */
+export function buttonEdit(title: string, callback: ClickHandler): UIElement {
+  return buttonSubtle("✎", title, callback);
+}
+/**
+ * Create a button that doesn't have the usual UI style for close and the like
+ */
+export function buttonSubtle(
+  icon: string,
+  title: string,
+  callback: ClickHandler
+): UIElement {
+  const button = document.createElement("span");
+  button.className = "close";
+  button.innerText = icon;
+  button.title = title;
+  button.style.cursor = "pointer";
+  button.addEventListener("click", (e) => {
+    e.stopPropagation();
+    callback(e);
+  });
+  return button;
+}
+
+/**
+ * Remove all child nodes from an element
+ */
+export function clearChildren(container: HTMLElement) {
+  while (container.hasChildNodes()) {
+    container.removeChild(container.lastChild!);
+  }
+}
+
+/**
+ * A collapsible section
+ * @param title the name of the section
+ * @param inner the contents of the section
+ */
+export function collapse(title: string, ...inner: UIElement[]): UIElement {
```

Comment:
this creates a collapsible section UIElement, but 'collapse()' as a function sounds like an action. Could it be something like 'collapsible()' instead?

Reply:
Sure. Fixed.